sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -117,6 +118,57 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     }
   }
 
+  override def visitCompoundOrSingleStatement(
+      ctx: CompoundOrSingleStatementContext): CompoundBody = {
+    Option(ctx.singleCompoundStatement()).map { s =>
+      visit(s).asInstanceOf[CompoundBody]
+    }.getOrElse {
+      val logicalPlan = visitSingleStatement(ctx.singleStatement())
+      CompoundBody(Seq(SingleStatement(
+        parsedPlan = logicalPlan,
+        sourceStart = ctx.start.getStartIndex,
+        sourceEnd = ctx.stop.getStopIndex + 1)))
+    }
+  }
+
+  override def visitSingleCompoundStatement(ctx: SingleCompoundStatementContext): CompoundBody = {
+    visit(ctx.beginEndCompoundBlock()).asInstanceOf[CompoundBody]
+  }
+
+  private def visitCompoundBodyImpl(ctx: CompoundBodyContext): CompoundBody = {
+    val buff = ListBuffer[CompoundPlanStatement]()
+    for (i <- 0 until ctx.getChildCount) {
+      val child = ctx.getChild(i)
+      val visitedChild = visit(ctx.getChild(i))
+      visitedChild match {
+        case statement: CompoundPlanStatement => buff += statement
+        case null if child.isInstanceOf[TerminalNodeImpl] =>
+      }
+    }
+
+    CompoundBody(buff.toSeq)
+  }
+
+  override def visitBeginEndCompoundBlock(ctx: BeginEndCompoundBlockContext): CompoundBody = {
+    visitCompoundBodyImpl(ctx.compoundBody())
+  }
+
+  override def visitCompoundBody(ctx: CompoundBodyContext): CompoundBody = {
+    visitCompoundBodyImpl(ctx)
+  }
+
+  override def visitCompoundStatement(ctx: CompoundStatementContext): CompoundPlanStatement = {
+    val child = visit(ctx.getChild(0))
```

Comment:
this looks very weird, do you mean `ctx.statement() != null`?

Reply:
The parser rule is
```
compoundStatement
    : statement
    | beginEndCompoundBlock
    ;
```
I think it should be very clear which one is matched.

Reply:
Done.